skbio/diversity/beta/_unifrac.py

```diff
@@ -0,0 +1,203 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2013--, scikit-bio development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from __future__ import absolute_import, division, print_function
+
+from skbio.util._decorator import experimental
+from skbio.diversity._base import (_validate_counts_vectors,
+                                   _validate_otu_ids_and_tree)
+
+
+def _observed_otu_counts(counts, otu_ids):
+    return {o: c for o, c in zip(otu_ids, counts) if c >= 1}
+
+
+def _validate(u_counts, v_counts, otu_ids, tree):
+    _validate_counts_vectors(u_counts=u_counts, v_counts=v_counts,
+                             suppress_cast=True)
+    _validate_otu_ids_and_tree(counts=u_counts, otu_ids=otu_ids, tree=tree)
+
+
+@experimental(as_of="0.4.0-dev")
+def unweighted_unifrac(u_counts, v_counts, otu_ids, tree,
```

Comment:
The Fast UniFrac code gets a lot of benefit from having its data structure cached, but that comes at a large expense on each call to the implementation. The present API doesn't allow for passing in an existing cached data structure. One option is to pack a cached vectorized version of the tree into `tree`. Another option would be to expand the API to allow `**kwargs`. I think the latter is better as it is more explicit, and it is possible that other implementations have have similar needs for being able to pass in additional information. Thoughts?